Minimal checkpointing support

adrianreber opened the pull request:
#### What type of PR is this?

/kind feature
/kind api-change

#### What this PR does / why we need it:

This PR introduces the minimal checkpoint support as discussed in https://github.com/kubernetes/enhancements/pull/1990

With the changes in the PR it is only possible to checkpoint a container. One of the main use cases, as discussed in the KEP of this PR is the forensic use case: Checkpoint a container and leave it running, take the checkpoint image and restart the container somewhere else (sandboxed environment) for analysis. 

Possible steps could be:
1. Checkpoint container out of Kubernetes: `curl -skv -X POST "https://localhost:10250/checkpoint/default/counters/wildfly"`
2. Take the checkpoint archive from `/var/lib/kubelet/checkpoints` and transfer it to another system
3. Use `crictl restore --import=<archive>` to restore a copy of the container somewhere else.

In the KEP we mentioned that checkpoints will be stored in OCI images and can be pushed to a registry. This is still planned but currently not implemented on the container engine level. It will be implemented in the future.

#### Which issue(s) this PR fixes:
This does not really completely fix any open issues, but it is a step forward to close issues like:

 - https://github.com/kubernetes/kubernetes/issues/3949
 - https://github.com/kubernetes/enhancements/pull/1990
 - https://github.com/kubernetes/enhancements/issues/2008

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
```release-note
Added a new feature gate `CheckpointRestore` to enable support to checkpoint containers. If enabled it is possible to checkpoint a container using the newly kubelet API (/checkpoint/{podNamespace}/{podName}/{containerName}).
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

 - https://github.com/kubernetes/enhancements/pull/1990
 - https://github.com/kubernetes/enhancements/issues/2008
 - https://github.com/kubernetes/kubernetes/pull/97194
 - https://github.com/kubernetes-sigs/cri-tools/pull/662
 - https://github.com/cri-o/cri-o/pull/4199
 - https://github.com/kubernetes/website/pull/31753

adrianreber committed:
kubelet: add CheckpointContainer() on the service level

Signed-off-by: Adrian Reber <areber@redhat.com>

adrianreber committed:
kubelet: add CheckpointContainer() to the runtime

Signed-off-by: Adrian Reber <areber@redhat.com>

adrianreber committed:
Add CRI API definitions for checkpoint/restore

This adds the minimal CRI API changes to support checkpoint/restore and
to enable implementation of these checkpoint/restore interfaces on
container engine level.

Signed-off-by: Adrian Reber <areber@redhat.com>

adrianreber committed:
make update after CRI API changes (checkpoint/restore)

Signed-off-by: Adrian Reber <areber@redhat.com>

adrianreber committed:
kubelet: add checkpoint/restore infrastructure

This adds the first infrastructure code parts to the kubelet
to support checkpoint/restore.

Signed-off-by: Adrian Reber <areber@redhat.com>

adrianreber committed:
kubelet: add CheckpointContainer in remote runtime

This is the first step to implement checkpointing and restoring of
container and containers starting from the lowest layer in the kubelet.

Signed-off-by: Adrian Reber <areber@redhat.com>

adrianreber committed:
kubelet: wire checkpoint container support through

This adds the last pieces to wire through the container checkpoint
support in the kubelet.

Signed-off-by: Adrian Reber <areber@redhat.com>

ehashman commented:
This needs to be implemented behind a feature flag.

/ok-to-test

adrianreber commented:
/test pull-kubernetes-unit

adrianreber commented:
Added unit tests and a feature flag (disabled by default).

ehashman commented:
> Does this PR introduce a user-facing change?
>
> NONE

A new feature is definitely a user-facing change, please add a release note.

adrianreber commented:
> > Does this PR introduce a user-facing change?
> > NONE
> 
> A new feature is definitely a user-facing change, please add a release note.

Right. Added. Initially, without the feature gate and being a kubelet only API, I thought it is not a user facing change. But with the feature gate it definitely is. Thanks for the reminder.

adrianreber commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

riteshnaik commented on pkg/kubelet/server/server.go:
```diff
@@ -878,6 +890,39 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+
+	found := false
+	for _, container := range pod.Spec.Containers {
+		if container.Name == containerName {
```
Relying on just featureGate seems like a hammer flag to enable checkpoints for all containers running inside the cluster. Should there be an additional container level spec flag to enable checkpoints for containers of interest and default being false? Without a container level spec flag and with just featureGate enabled, all containers would have a specific checkpoint on the node and depending on number of containers running on the node it might create a disk pressure if enough disk space is not allocated. Any thoughts?

riteshnaik commented on pkg/kubelet/server/server.go:
```diff
@@ -878,6 +890,39 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+// checkpoint handles the checkpoint API request. It checks if the requested
+// podNamespace, pod and container actually exist and only then calls out
+// to the runtime to actually checkpoint the container.
+func (s *Server) checkpoint(request *restful.Request, response *restful.Response) {
```
I might have missed it or maybe not part of this PR, when is checkpoint triggered in the container lifecycle? Is it when the container or pod is ready? is it configurable? It would be helpful to trigger the checkpoint if the container is in a certain state.

adrianreber replied:
The kubelet API is the only way to trigger a checkpoint, yes.

riteshnaik replied:
The following question is out of scope for this PR but just picking your brain for future work. 

* Could the trigger to checkpoint container or pod of interest be specified through "container spec" for container checkpoint and "pod spec" for pod checkpoint and at the particular state of the container lifecycle?

derekwaynecarr replied:
initially we want to avoid exposing checkpoint operations from the k8s control plane itself.

we need to better understand the pros/cons from a forensic use case.

riteshnaik commented on pkg/kubelet/server/server.go:
```diff
@@ -878,6 +890,39 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+	containerName := request.PathParameter("containerName")
+
+	found := false
+	for _, container := range pod.Spec.Containers {
```
It might or might not be necessary to have the checkpoint of every container if there are multiple replicas of the same container. Not sure, if it would scale well if there are 1000s of replicas of containers and corresponding checkpoints for each replica. This would further occupy space on disk if local or registry if remote. Maybe an option to just have 1 checkpoint for each replica set would be handy.  Any thoughts?

riteshnaik commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:
```diff
@@ -1502,3 +1507,70 @@ message ReopenContainerLogRequest {
+
+// Common options used for checkpointing and restoring.
+message CheckpointRestoreOptions {
+    // Keep temporary files. Like log files. Helpful for debugging.
```
What would it look like to add additional options or override existing ones? One prerequisite for checkpoint/restore to work is for CRIU to be available and so options in the /etc/criu/default.conf would be honored? But one drawback with this would be that it would apply to all containers and options might differ depending on the specific container. Should this be made part of container spec somehow? Any thoughts on overall CRIU configurations for specific container?

adrianreber replied:
This questions feels out of scope of this PR. For the current use case and the goal of this PR (minimal checkpoint support and CRI API changes to start implementation on the container engine level) I do not think it is necessary to already have a clear definition how to handle additional parameters.

There is, however, already a runc annotation which supports specifying a CRIU configuration file in the container's `config.json` file. If we enable this in the container engine level, to specify per container CRIU options we can extend the CRI API, if necessary, to pass container specific CRIU options from the kubelet to the container engine via the CRI API.

riteshnaik replied:
I like the idea of extending CRI API to specify per container CRIU options from kubelet to the container engine and I can understand this being out of this PR. Is there a way to track these thoughts/ideas somewhere or maybe this could be a 1 liner comment so that we don't lose track of them?

riteshnaik commented on pkg/features/kube_features.go:
```diff
@@ -765,6 +765,12 @@ const (
+	// alpha: v1.23
+	//
+	// Enables Checkpoint/Restore in the kubelet
+	CheckpointRestore featuregate.Feature = "CheckpointRestore"
```
Would this be confusing from the user's point of view? It's fine for long-term vision but for now, just checkpoint would be enabled with this PR but reading this seems like both Checkpoint and Restore functionalities are enabled.

adrianreber replied:
I also thought about it. Not sure. The long term goal is to also have restore support. Not sure if it makes sense to have two feature gates if we ever introduce restore support in the future. I opted for a combined feature gate name as it will be unlikely to have only checkpoint or only restore enabled. The corresponding documentation (does not yet exist) should help make it clear what the feature gate actually means. In the end it is just a string and I am open to changing it. The current name is just what made most sense to me.

riteshnaik commented on pkg/kubelet/cri/remote/fake/fake_runtime.go:
```diff
@@ -322,3 +322,18 @@ func (f *RemoteRuntime) ReopenContainerLog(ctx context.Context, req *kubeapi.Reo
+	return &kubeapi.CheckpointContainerResponse{}, nil
+}
+
+// CheckpointContainer checkpoints the given container.
```
Is this method needed for this PR? and if yes, the comment here needs to be updated to describe the restore

adrianreber replied:
Yes, it is needed because of the CRI API changes and it does not compile without this. Thanks for pointing out the wrong comment. Fixed.

adrianreber commented:
Something strange happened with the review comments from @riteshnaik . It seems a few of the comments disappeared as well as my answers to it. Not sure what happened.

adrianreber commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

/test pull-kubernetes-integration

adrianreber commented:
/test pull-kubernetes-e2e-gce-alpha-features

adrianreber commented:
/test pull-kubernetes-e2e-gce-alpha-features

dims commented:
Is this PR still needed, please rebase if so (or we can close it?)

adrianreber commented:
/test pull-kubernetes-unit

adrianreber committed:
Add e2e tests for checkpointing

Signed-off-by: Adrian Reber <areber@redhat.com>

adrianreber commented:
/test pull-kubernetes-conformance-kind-ipv6-parallel

adrianreber commented:
/test pull-kubernetes-e2e-kind-ipv6

adrianreber commented:
/test pull-kubernetes-e2e-kind-ipv6

adrianreber commented:
The corresponding KEP (https://github.com/kubernetes/enhancements/pull/1990) has been merged and this PR has been updated based on the reviews from the KEP to include e2e tests.

Ready for reviews.

adrianreber commented:
@SergeyKanzhelev I see

> [SergeyKanzhelev](https://github.com/SergeyKanzhelev) moved this from Triage to Archive-it in [SIG Node CI/Test Board](https://github.com/orgs/kubernetes/projects/43#card-77085451) [3 days ago](https://github.com/kubernetes/kubernetes/pull/104907#event-5997797109)

Archive-it sounds strange as this PR is actually waiting for reviewers, Not sure if I misunderstood your change of it is not correct.

SergeyKanzhelev commented:
CI/Test board is used for test related PRs. Instead of removing from the board we are archiving it as otherwise PRs will show up over and over - we are looking which PRs to add to the board based on labels and any PR that has any test code is marked as test-related.

adrianreber commented:
/test pull-kubernetes-conformance-kind-ipv6-parallel

adrianreber commented:
Rebased and force pushed. The doc pull request is now also ready at https://github.com/kubernetes/website/pull/31753 and waiting on this code PR to get merged.

Reviewers PTAL

adrianreber commented:
/test pull-kubernetes-unit

adrianreber commented:
/test pull-kubernetes-unit

adrianreber commented:
/retest-required

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1512,3 +1517,70 @@ message ReopenContainerLogRequest {
+// Common options used for checkpointing and restoring.
+message CheckpointRestoreOptions {
+    // Keep temporary files. Like log files. Helpful for debugging.
+    bool keep = 1;
```
Can this be named `keep_temporary_files`?

adrianreber replied:
done

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1512,3 +1517,70 @@ message ReopenContainerLogRequest {
+    // Keep temporary files. Like log files. Helpful for debugging.
+    bool keep = 1;
+    // Checkpoint/Restore the container with established TCP connections.
+    bool tcp_established = 2;
```
Maybe `keep_network_state`?

adrianreber replied:
done

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1512,3 +1517,70 @@ message ReopenContainerLogRequest {
+    // Checkpoint/Restore the container with established TCP connections.
+    bool tcp_established = 2;
+    // Location of the checkpoint archive used for export/import
+    string archive = 3;
```
`archive_location`

adrianreber replied:
done

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1512,3 +1517,70 @@ message ReopenContainerLogRequest {
+message CheckpointContainerResponse {
+}
+
+message RestoreContainerOptions {
```
I think we could probably leave Restore out for now.

adrianreber replied:
Thanks for the review. I think we already briefly discussed this in the KEP and it is mentioned there also. My argument for keeping the `Restore` part in the CRI API it is necessary to restore checkpoints in CRI implementations like CRI-O via crictl.

https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2008-forensic-container-checkpointing/README.md?plain=1#L111

adrianreber replied:
> do you have any proposal(s) for new crictl commands

https://github.com/kubernetes-sigs/cri-tools/pull/662

> I may be way off base.. what is the expected state of the restored container?

Exactly what it was during checkpointing.  What do you mean by state?

mikebrow replied:
for forensics I expected the concept of container state to be one of https://github.com/opencontainers/runc/blob/51e607f2cdd8ccf7f39f29aab531d674e487b311/libcontainer/container.go#L29-L42

The docs for runc checkpoint mention .. checkpoint for a "running container"  Does that mean this checkpointing feature will only work if the container is in a running state?

mikebrow replied:
in my head there is a run container state matrix for checkpoint and another for restore.. and I don't know how to fill it out exactly.. Apologies, in this review I have not yet tried to build the parts and run it locally.

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1512,3 +1517,70 @@ message ReopenContainerLogRequest {
+}
+
+message CheckpointContainerRequest {
+    // ID of the container or pod to be checkpointed.
```
Should we have separate requests for pod vs. containers? 
Pods may be constructed differently e.g. they could be pinned namespaces or pause (infra) containers or VMs.

adrianreber replied:
> Should we have separate requests for pod vs. containers?

From what I have tested so far with pods and containers it works well passing the ID (container or pod) to CRI-O. What I am doing in CRI-O is trying the ID first as a container ID and then as a POD ID (or the other way around) and if one matches CRI-O will checkpoint the container or pod.

> Pods may be constructed differently e.g. they could be pinned namespaces or pause (infra) containers or VMs.

Yes. The resulting checkpoints are different but CRI-O already knows how to checkpoint a container in contrast to a pod. From my point of view it can be done using a single parameter. We can also split it up in a container id and pod id parameter. It did not seem necessary so far.

adrianreber replied:
> The container id should be restricted to just container ids, overloading container ids and pod ids would require api changes.

Sorry, I also do not understand this comment. Currently I tell CRI-O to checkpoint the thing with the given ID and CRI-O figures out if it is a pod or a container.

Also, why would it require API changes? This PR introduces the API. Sorry, but I do not understand what you exactly mean. If it should be split, that is no problem, I just do not understand the reason. I guess.

mikebrow replied:
There is no rule in CRI that requires pod and container ids to be drawn from the same bucket and have a status associated to indicate whether the id is a pod or container. If you check the CRI apis you will see they either receive a pod id or a container id, never a pod or container id. 

The CRI api supports shortcutting ids, via a lookup. In your example, container ids b0c15999fad13 
052ff0835b48d, in pod 7bc90f445a4bc... I can stop the pod using crictl stopp 7, i can stop the first container using stop b.  I can't stop the pod using stop 7, or stop 7bc90f445a4bc.

adrianreber replied:
I see. Would you recommend to have completely different CRI API calls to do checkpointing and restoring?

I have the following RPC call in my PR:
```
rpc CheckpointContainer(CheckpointContainerRequest) returns (CheckpointContainerResponse) {}
```

Should that handle pod checkpointing in the future using an additional parameter? Should I rename it to have a more generic name so that it makes more sense in the future to include pod checkpointing? Should I just leave it as it is and use another RPC call in the future if we ever add pod checkpointing?

mikebrow replied:
Further .. though the odds are immense there could be a point in time where there is a pod id and a container id that have the same value yet are unique because one is a pod and the other is a container, if that is not possible, or if it should not be possible, we probably need to specify that somewhere.

adrianreber commented:
@mrunalp Thanks for your review. I did some changes as suggested. I tried to answer your questions around `Restore` in the CRI API and shared pod and container ID parameter.

elchead commented:
@adrianreber I'm trying to set up a cluster with your feature and am not sure about what I need. Besides building your kubernetes fork for the new `kubelet` and `kube-apiserver`, I also need to install your CRI-O PR correct?

adrianreber commented:
> @adrianreber I'm trying to set up a cluster with your feature and am not sure about what I need. Besides building your kubernetes fork for the new `kubelet` and `kube-apiserver`, I also need to install your CRI-O PR correct?

Yes, the CRI-O changes are needed. The CRI-O pull request currently does not have the CRI API changes to make CRI-O CI happy. So currently it requires additional code changes to actually test it end-to-end.

The changes on top of the existing CRI-O PR are minimal. I can help with it. It is basically wiring through the CRI-API to the corresponding CRI-O functions (which are part of my PR). Let me know and I can create a second PR for you to try.

elchead commented:
> 

thanks for your offer to help! It would be great if you could help me to make the necessary changes on the CRI-O PR. I'm also on discord (elchead#1904) if this would help with communication :)

adrianreber commented:
> > > @adrianreber I'm trying to set up a cluster with your feature and am not sure about what I need. Besides building your kubernetes fork for the new `kubelet` and `kube-apiserver`, I also need to install your CRI-O PR correct?
> > 
> > 
> > Yes, the CRI-O changes are needed. The CRI-O pull request currently does not have the CRI API changes to make CRI-O CI happy. So currently it requires additional code changes to actually test it end-to-end.
> > The changes on top of the existing CRI-O PR are minimal. I can help with it. It is basically wiring through the CRI-API to the corresponding CRI-O functions (which are part of my PR). Let me know and I can create a second PR for you to try.
> 
> @adrianreber thanks for your offer to help! It would be great if you could help me to make the necessary changes on the CRI-O PR.

I will let you know once it is ready.

> I'm also on discord (elchead#1904) if this would help with communication :)

I am on Matrix and Slack :wink:

elchead commented:
> 

I guess slack only works through invitation outside the organization, but you could invite me here (adrian.stobbe@sap.com) :)

https://slack.com/help/articles/1500001422062-Use-Slack-Connect-to-start-a-DM-with-someone-at-another-company#:~:text=You%20can%20use%20Slack%20Connect,must%20be%20sent%20and%20accepted.

adrianreber commented:
@elchead A CRI-O branch with the CRI API wired through can be found here: https://github.com/adrianreber/cri-o/tree/checkpoint-restore-support-cri-api

mikebrow commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1alpha2/api.proto:
```diff
@@ -1512,3 +1517,70 @@ message ReopenContainerLogRequest {
+    // ID of the restored container or pod
+    string id = 1;
+    // True if the restored container/pod is a pod
+    bool pod = 2;
```
see prior comment regarding this should only be a container id

adrianreber replied:
Hmm. The idea was something else. This also feels like I am not understanding what you mean.

The restore answer tells me if I restored a pod with multiple containers or just a single container. Using the boolean `pod` the client has the possibility to see if a pod with multiple containers or just a single container is restored.

adrianreber replied:
Removed anything about pods in the latest version of the PR. Only the destination Pod of the restore is still left as part of the CRI.

mikebrow reviewed (commented):
See comments..

General comment, the CRI runtime services are very pod focused and because this version of the KEP is focused just on container checkpoints I'm having a hard time understanding the potential use cases of the partial implementation. For example, I'm trying to grok how someone would actually do forensics on the checkpointed container. Do you have a general user level case description of how this will work via the crictl tool over the cri api in a "forensics" context? Or is this mostly focused on using dump/debug tools against the tar?

adrianreber commented:
> General comment, the CRI runtime services are very pod focused and because this version of the KEP is focused just on container checkpoints I'm having a hard time understanding the potential use cases of the partial implementation. For example, I'm trying to grok how someone would actually do forensics on the checkpointed container. Do you have a general user level case description of how this will work via the crictl tool over the cri api in a "forensics" context? Or is this mostly focused on using dump/debug tools against the tar?

@mikebrow Thanks for your review. Checkpointed container can be restored using `crictl -i path/to/checkpoint-archive.tar --pod=$POD_ID`. This questions feels more like a KEP level question, so not sure how to answer it better than, yes the corresponding `crictl` patches are also ready and waiting on the CRI API changes to be merged.

adrianreber commented on pkg/kubelet/cri/remote/remote_runtime.go:
```diff
@@ -1149,3 +1150,82 @@ func (r *remoteRuntimeService) ReopenContainerLog(containerID string) (err error
+			// The CRI implementation reads the ID and checkpoints
+			// a container if the ID represents a container or it
+			// checkpoints a Pod if the ID represents a Pod.
+			&runtimeapiV1alpha2.CheckpointContainerRequest{
```
> at some point we should stop adding new apis to old versions of the api.. :-)

I am just following the existing code and it currently seems to support v1 and v1alpha2.

> Maybe add a check if returned error v1 and v1alpha2 responds that the runtime does not include this new CheckpointContainer service.. Or is the grpc error clean enough output?

I am not sure I understand this comment correctly. Why should there be a check? Can you clarify what you meant.

mikebrow replied:
> I am just following the existing code and it currently seems to support v1 and v1alpha2.

We should now be treating v1alpha2 as frozen. Confirm with @mrunalp. This, at least because, older versions of container runtimes implementing the v1alpha2 api will not have your new service api.

> > Maybe add a check if returned error v1 and v1alpha2 responds that the runtime does not include this new CheckpointContainer service.. Or is the grpc error clean enough output?
> 
> I am not sure I understand this comment correctly. Why should there be a check? Can you clarify what you meant.

Older versions of container runtimes that do not have the new API will cause you to get a GRPC error here.

adrianreber replied:
I just tried it without the checkpoint CRI call implemented and I get a 500 with following text: "checkpointing of counter failed (container counter not found)"

That is not correct. I will look into getting the correct error message to the user. Good point.

adrianreber replied:
@mrunalp Should I remove everything concerning v1alpha2 from this PR?

mrunalp replied:
Yes, we shouldn't add features to v1alpha2.

adrianreber commented on pkg/kubelet/kubelet.go:
```diff
@@ -2411,6 +2415,22 @@ func (kl *Kubelet) fastStatusUpdateOnce() {
+		return fmt.Errorf("container %v not found", containerName)
+	}
+
+	if err := kl.containerRuntime.CheckpointContainer(string(container.ID.ID), kl.getCheckpointsDir()); err != nil {
```
> For example, here I expected that this path would include something like the pod namespace and/or pod id.

The path for the resulting checkpoint archive location? pod namespace and/or pod id could be added, but the container ID is unique, so there will be a way to figure out the origin of the checkpointed container. If this is what you meant it can be added if necessary. The plan for the checkpoint request is to pass additional information about the checkpoint creation via POST, for this PR I tried to keep it really simple as additional parameters can be added later.

mikebrow replied:
> The path for the resulting checkpoint archive location? pod namespace and/or pod id could be added, but the container ID is unique, so there will be a way to figure out the origin of the checkpointed container. 

It's unique but can be reused. The pod may be gone by the time the user goes to look for the checkpoint file.

> The plan for the checkpoint request is to pass additional information about the checkpoint creation via POST, for this PR I tried to keep it really simple as additional parameters can be added later.

right and we can also add more context later, and align that with the context needed for pod checkpointing

adrianreber replied:
Right, makes sense. I will change the name of the checkpoint archive.

adrianreber commented:
As this questions came up a couple of times I am including all the currently necessary steps to checkpoint and restore a container. This PR is only about checkpointing but it also adds restore to the CRI API and the following is how I can currently use it:
```
### Create a pod with two containers
# cluster/kubectl.sh apply -f two.yaml 
pod/counters created
### Show running containers
# crictl ps -a
CONTAINER           IMAGE                                                                                                       CREATED                  STATE               NAME                ATTEMPT             POD ID
b0c15999fad13       quay.io/adrianreber/counter@sha256:390d7063f76d97d42e74b3ac826bdd453dec70c9df4614574d661664e6bd2c34         Less than a second ago   Running             counter             0                   7bc90f445a4bc
052ff0835b48d       quay.io/adrianreber/wildfly-hello@sha256:225d458da20231fe6843bde19d0c4ade04f18143a82b297962d23676bebaa42d   3 seconds ago            Running             wildfly             0                   7bc90f445a4bc
### Access one of the the containers
# curl `crictl inspect b0c15999fad13 | jq '.info.runtimeSpec.annotations."io.kubernetes.cri-o.IP.0"' -r`:8088
counter: 0
# curl `crictl inspect b0c15999fad13 | jq '.info.runtimeSpec.annotations."io.kubernetes.cri-o.IP.0"' -r`:8088
counter: 1
### Checkpoint one of the containers
# curl -sk -X POST "https://localhost:10250/checkpoint/default/counters/counter"
### Create a new pod
# POD_ID=$(crictl runp pod-config.json )
### Restore the checkpointed container into that new pod
# crictl restore -i /var/lib/kubelet/checkpoints/b0c15999fad13ac730336bdab6e8ff7142abfae4241f940c0c0ba0671e93265b.tar --pod=$POD_ID 
[
  {
    "id": "3140ccceac3edf9b330095eaee692265730341f157864bc38d5ecfa1e9b5561a",
    "restored_containers": [
      "186ecbdf7dd249a34df53f3670055a7ad0a4aa8e919ee2a0cbc21b7e694a4745"
    ]
  }
]
### I get the information about the destination pod and the IDs of the restored containers back
### Access the restored container
# curl `crictl inspect 186ecbdf7dd249a34df53f3670055a7ad0a4aa8e919ee2a0cbc21b7e694a4745 | jq '.info.runtimeSpec.annotations."io.kubernetes.cri-o.IP.0"' -r`:8088
counter: 2
### List running container. Original container and restored containers
# crictl ps
CONTAINER           IMAGE                                                                                                       CREATED             STATE               NAME                ATTEMPT             POD ID
186ecbdf7dd24       quay.io/adrianreber/counter:latest                                                                          9 minutes ago       Running             counter             0                   3140ccceac3ed
b0c15999fad13       quay.io/adrianreber/counter@sha256:390d7063f76d97d42e74b3ac826bdd453dec70c9df4614574d661664e6bd2c34         10 minutes ago      Running             counter             0                   7bc90f445a4bc
052ff0835b48d       quay.io/adrianreber/wildfly-hello@sha256:225d458da20231fe6843bde19d0c4ade04f18143a82b297962d23676bebaa42d   10 minutes ago      Running             wildfly             0                   7bc90f445a4bc
```

adrianreber commented on pkg/kubelet/cri/remote/remote_runtime.go:
```diff
@@ -1149,3 +1150,82 @@ func (r *remoteRuntimeService) ReopenContainerLog(containerID string) (err error
+							// the CRI API already contains the necessary field
+							// to tell the CRI implementation which compression
+							// to use.
+							ArchiveLocation: filepath.Join(checkpointDir, containerID+".tar"),
```
> is this the right place to check if exists and start rolling a counter for the nth checkpoint

Yes, probably. The goal was to be as simple as possible and add any additional features in later versions. But if wanted I can add it already here and now.

mikebrow replied:
I think so. That or throw an error if already exists, or document to the user that they should move or delete ASAP.

adrianreber commented on pkg/kubelet/kubelet.go:
```diff
@@ -2411,6 +2415,22 @@ func (kl *Kubelet) fastStatusUpdateOnce() {
 	}
 }
 
+func (kl *Kubelet) CheckpointContainer(podUID types.UID, podFullName, containerName string) error {
```
added

adrianreber commented on pkg/kubelet/cri/remote/remote_runtime.go:
```diff
@@ -1149,3 +1150,82 @@ func (r *remoteRuntimeService) ReopenContainerLog(containerID string) (err error
 	return nil
 }
+
+func (r *remoteRuntimeService) CheckpointContainer(containerID, checkpointDir string) error {
```
added

adrianreber commented on pkg/kubelet/cri/remote/remote_runtime.go:
```diff
@@ -1149,3 +1150,82 @@ func (r *remoteRuntimeService) ReopenContainerLog(containerID string) (err error
+					Id: containerID,
+					Options: &runtimeapi.CheckpointContainerOptions{
+						CommonOptions: &runtimeapi.CheckpointRestoreOptions{
+							// Currently the archive is not yet compressed, but
```
added

adrianreber commented:
@mikebrow I tried to address your questions and changed the PR as suggested.

A couple of your review comments were not clear to me and I asked for clarifications. Over the last 1.5 years there have been different PRs using the proposed changes to checkpoint and restore complete pods and it has been simplified over time to the now merged KEP to just be about container checkpointing. This PR is based on the KEP and just trying to implement what has been agreed upon in the KEP. Even if this PR is very limited in its functionality that is as discussed in the KEP.

Let me know if you have any further questions or comments.

adrianreber commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

adrianreber commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

mikebrow commented:
> @mikebrow I tried to address your questions and changed the PR as suggested.
> 
> A couple of your review comments were not clear to me and I asked for clarifications. Over the last 1.5 years there have been different PRs using the proposed changes to checkpoint and restore complete pods and it has been simplified over time to the now merged KEP to just be about container checkpointing. This PR is based on the KEP and just trying to implement what has been agreed upon in the KEP. Even if this PR is very limited in its functionality that is as discussed in the KEP.
> 
> Let me know if you have any further questions or comments.

nod.. yes I read the KEP a couple times while reviewing :-)

adrianreber commented:
With the current github problems all tests are failing, but I will restart them later.

@mikebrow I think I followed most of your suggestions and changed the PR accordingly:

* The checkpoint image will no longer be overwritten. There is now a counter appended to the checkpoint archive name.
* The checkpoint is no longer name `<containerid>.tar`, but `checkpoint-<podname>-<containername>-<counter>.tar`
* It is possible to pass a custom checkpoint archive name (it is not possible to change the destination directory only the name of the actual file>
* I removed everything concerning pods from the CRI API
* I checked that the output from the POST request has better error messages:
   * ` 500 checkpointing of counter failed (rpc error: code = Unimplemented desc = unknown method CheckpointContainer for service runtime.v1.RuntimeService)`
   * `500 checkpointing of counter failed (CheckpointContainer is only supported in the CRI v1 runtime API)`
   * `200 /var/lib/kubelet/checkpoints/checkpoint-counters_default-counter-4.tar`

Thanks for so far for your feedback. Let me know if you have any further questions/suggestions.

adrianreber commented:
/retest

adrianreber commented:
/retest

adrianreber commented:
/test pull-kubernetes-conformance-kind-ipv6-parallel

adrianreber commented:
/test pull-kubernetes-integration

ehashman commented:
/priority important-soon
/milestone v1.24
/triage accepted
/label api-review
/assign

ehashman commented on test/e2e/common/node/checkpoint_container.go:
```diff
@@ -0,0 +1,157 @@
+	}
+}
+
+var _ = SIGDescribe("Checkpoint Container [NodeFeature:CheckpointContainer]", func() {
```
`[NodeFeature:*]` only work on node_e2e tests and not tests in the `test/e2e/...` directory. I don't think this test needs a full cluster, please move this file to `test/e2e_node/` and then we'll be able to pick it up with the `pull-kubernetes-node-e2e-containerd-features` or `pull-kubernetes-node-crio-e2e` jobs.

adrianreber replied:
No problem. Moved to `test/e2e_node/`.

ehashman commented on pkg/kubelet/server/server_test.go:
```diff
@@ -546,6 +556,9 @@ func TestAuthzCoverage(t *testing.T) {
 
 func TestAuthFilters(t *testing.T) {
+	// Enable features.ContainerCheckpoint during test
+	defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.ContainerCheckpoint, true)()
```
Can we also test this with the feature flag enabled? Does any behaviour change?

adrianreber replied:
Do you mean with the feature flag **disabled**? Maybe I am using `SetFeatureGateDuringTest()` wrongly. I am currently confused why it is using `defer`.

I am running the tests like this: `KUBE_TEST_ARGS="-v -failfast" bash -x hack/make-rules/test.sh ./pkg/kubelet/server` and without the defer line I get following error:
```
=== RUN   TestAuthFilters/POST:/checkpoint/{podNamespace}/{podName}/{containerName}
    server_test.go:599: 
        	Error Trace:	server_test.go:599
        	Error:      	Not equal: 
        	            	expected: 403
        	            	actual  : 404
        	Test:       	TestAuthFilters/POST:/checkpoint/{podNamespace}/{podName}/{containerName}
    server_test.go:600: 
        	Error Trace:	server_test.go:600
        	Error:      	Should be true
        	Test:       	TestAuthFilters/POST:/checkpoint/{podNamespace}/{podName}/{containerName}
        	Messages:   	Authenticate was not called
    server_test.go:601: 
        	Error Trace:	server_test.go:601
        	Error:      	Should be true
        	Test:       	TestAuthFilters/POST:/checkpoint/{podNamespace}/{podName}/{containerName}
        	Messages:   	Attributes were not called
    server_test.go:602: 
        	Error Trace:	server_test.go:602
        	Error:      	Should be true
        	Test:       	TestAuthFilters/POST:/checkpoint/{podNamespace}/{podName}/{containerName}
        	Messages:   	Authorize was not called
--- FAIL: TestAuthFilters (0.26s)
```
With the defer line the test passes. My idea was to enable the feature during this test to make the test pass, but I am no longer sure that makes sense. Any recommendations from your side how to use it correctly?

endocrimes replied:
The current behaviour here is correct - featuregatetesting.SetFeatureGateDuringTest(...) will immediately set the feature gate, and returns a function to revert the feature gate to its original value. This is the function that is then `defer`'d

ehashman commented on test/e2e/common/node/checkpoint_container.go:
```diff
@@ -0,0 +1,157 @@
+			// we should get a 404. So a 404 is (also) a good sign.
+			if (int(statusError.ErrStatus.Code)) == http.StatusNotFound {
+				ginkgo.By("Feature 'ContainerCheckpoint' is not enabled and not available")
+				return
```
In this case, should we skip the test?

adrianreber replied:
Changed to `Skip()`.

ehashman commented on test/e2e/common/node/checkpoint_container.go:
```diff
@@ -0,0 +1,157 @@
+					statusError.ErrStatus.Message,
+					"(rpc error: code = Unimplemented desc = unknown method CheckpointContainer",
+				) {
+					ginkgo.By("Container engine does not implement 'CheckpointContainer'")
+					return
+				}
+			}
+			framework.Failf("Unexpected status code (%d) during 'CheckpointContainer'", statusError.ErrStatus.Code)
```
This seems like a fair thing to test given that container runtimes can't yet support the new CRI.

adrianreber replied:
Not sure if this is just a comment or a request to change something. Are you asking for a change? It does not sound like it.

ehashman replied:
No change, just a comment.

ehashman reviewed (commented):
/remove-label api-review

This is CRI changes only for the API change.

adrianreber commented:
/test?

adrianreber commented:
/test all

adrianreber commented:
/test pull-kubernetes-verify-govet-levee

adrianreber commented:
/retest

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1512,3 +1517,68 @@ message ReopenContainerLogRequest {
 }
+
+// Common options used for checkpointing and restoring.
+message CheckpointRestoreOptions {
```
My preference is to add as few options as we can get away with. We could choose
reasonable defaults for the rest. Do we need anything besides the archive_location 
to begin with? We can always add options later if needed.

adrianreber replied:
I removed everything beside `archive_location` and `archive_is_oci_image` to easily distinguish if the destination is a local image or if should be pushed to a registry.

mrunalp replied:
I think that we can have it be local and have a path configured in the runtime as the default. The transportation of the checkpointed archive to a registry is a separate concern out of the CRI, so we don't need that. It could be handled by some operator later on.

adrianreber replied:
Also removed `archive_is_oci_image`.

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1512,3 +1517,68 @@ message ReopenContainerLogRequest {
+    string name = 3;
+    // The labels and annotations are in the same form as in ContainerConfig
+    // during CreateContainerRequest.
+    map<string, string> labels = 4;
```
Is this needed or something we can add later?

adrianreber replied:
Not needed. That was a left-over from the end-to-end implementation of the drain use case from the beginning of the discussions.

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1512,3 +1517,68 @@ message ReopenContainerLogRequest {
+    // The labels and annotations are in the same form as in ContainerConfig
+    // during CreateContainerRequest.
+    map<string, string> labels = 4;
+    map<string, string> annotations = 5;
```
Same question for this.

adrianreber replied:
Not needed. That was a left-over from the end-to-end implementation of the drain use case from the beginning of the discussions.

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1512,3 +1517,68 @@ message ReopenContainerLogRequest {
+    // the pod ID.
+    // This tells the container engine to replace existing mount
+    // sources with a new source.
+    map<string, string> change_mounts = 6;
```
Can you give an example of how this may be used?

adrianreber replied:
That is needed to restore a container outside of Kubernetes. Especially for the forensic use case it is important.

Each container has a couple of Kubernetes specific bind mounts in the container. One example is `/etc/hosts`. In one of my examples it is bind mounted from `/var/lib/kubelet/pods/3f24e9ed-78e5-48ad-b300-0a10ba5bb181/etc-hosts`. During restore, on another node, potentially without Kubernetes CRIU will try to restore the bind mount from `/var/lib/kubelet/pods/3f24e9ed-78e5-48ad-b300-0a10ba5bb181/etc-hosts` to `/etc/hosts` and as the container external file is missing, the restore will fail.

With this option in the CRI I can tell the container engine via crictl to bind mount another file as `/etc/hosts` from the outside of the container.

I kept this one in, as it essential for the restore part outside of Kubernetes for the forensic use case.

All other currently unneeded CRI entries have been removed.

mrunalp replied:
@adrianreber Couldn't we add some hard-coded list here for well-known mount path patterns so we can do this ID replacement in the restore code in the runtime?

adrianreber replied:
hard-coded does not sound doable, but I removed it and I will provide a tool to adapt the checkpoint archive. That way we can change the mount points after checkpointing but before restoring. It will result in a manual extra step before being able to restore it. That way we can learn if it would be helpful/important to have it in the CRI API or if it works with an additional step.

adrianreber commented:
/retest

dims commented:
/assign @derekwaynecarr @mrunalp

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1514,3 +1519,56 @@ message ReopenContainerLogRequest {
+message CheckpointContainerOptions {
+    CheckpointRestoreOptions common_options = 1;
+    // Leave the container running after checkpointing
+    bool leave_running = 2;
```
This could be the default for the forensic use case so can be dropped.

adrianreber replied:
dropped

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1514,3 +1519,56 @@ message ReopenContainerLogRequest {
+}
+message RestoreContainerRequest {
+    // ID of the container to be restored.
+    string id = 1;
```
The runtimes pick an ID for containers, so we don't need this. The id in the response should be enough.

adrianreber replied:
dropped

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1514,3 +1519,56 @@ message ReopenContainerLogRequest {
+    // the pod ID.
+    // This tells the container engine to replace existing mount
+    // sources with a new source.
+    map<string, string> change_mounts = 4;
```
Could this be stored in the archive for now?

ehashman commented:
/test pull-kubernetes-node-e2e-containerd-features

adrianreber commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1514,3 +1519,56 @@ message ReopenContainerLogRequest {
+    string pod_sandbox_id = 2;
+    // Restoring multiple copies of a checkpointed container may
+    // require a new name for the restored container.
+    string name = 3;
```
dropped

adrianreber commented:
/test pull-kubernetes-node-e2e-containerd-features

rphillips commented on pkg/kubelet/server/server.go:
```diff
@@ -878,6 +891,52 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+			return
+		}
+	}
+	klog.InfoS("Request PAram login", "options", options)
```
this log line should be at a certain log level.

adrianreber replied:
I removed that line, that was a debug message left over and not necessary any more.

rphillips commented:
Besides the one log line nit... lgtm

adrianreber commented:
/test pull-kubernetes-node-e2e-containerd-features

adrianreber commented:
/test pull-kubernetes-node-e2e-containerd-features

adrianreber commented:
/retest

adrianreber commented:
/retest

adrianreber commented:
/retest

ehashman commented:
Test is running and getting skipped. Failures for `NodeFeature:HostAccess` tests are unrelated.

```
E2eNode Suite: [sig-node] Checkpoint Container [NodeFeature:CheckpointContainer] will checkpoint a container out of a pod

Reason: test/e2e_node/checkpoint_container.go:72 Feature 'ContainerCheckpoint' is not enabled and not available test/e2e_node/checkpoint_container.go:131
```

adrianreber commented:
> Test is running and getting skipped. Failures for `NodeFeature:HostAccess` tests are unrelated.
> 
> ```
> E2eNode Suite: [sig-node] Checkpoint Container [NodeFeature:CheckpointContainer] will checkpoint a container out of a pod
> 
> Reason: test/e2e_node/checkpoint_container.go:72 Feature 'ContainerCheckpoint' is not enabled and not available test/e2e_node/checkpoint_container.go:131
> ```

Is that an environment where the feature should be enabled? If it should be enabled then it should not be skipped.

ehashman commented:
@adrianreber would need all alpha gates on and run through node features. We might not have a presubmit for that. Submitted a PR to add one, but I don't think this is blocking.

rphillips commented:
/lgtm

gracenng commented:
hi folks this needs approve label for 1.24 code freeze

ehashman commented:
/milestone clear

missed code freeze

ehashman commented:
/test pull-kubernetes-node-e2e-containerd-alpha-features

adrianreber commented:
/test pull-kubernetes-node-e2e-containerd-alpha-features

arierev commented:
Is there a problem with merging this feature to 1.24? 

May be something that may be waiting an intervention (code fix, or test execution) of the developer?
If not - what is precluding this feature to be merged?

I am asking as we are (impatiently) waiting for this feature to be supported by Kubernetes, for a use case of expediting pod startup.
The application initialization is very long (many minutes) - this time can be drastically reduced by using the "restore" of a previously-taken checkpoint.  

WIthout this feature, using Kuberentes auto-scaling is practically useless for our application, as until the the process in a newly-launched Pod is ready to serve, the conditions that triggered that scale-out have changed so much that this scale-out is too-little-too-late or not needed any more....

Thanks!

adrianreber commented:
/test pull-kubernetes-node-e2e-containerd-alpha-features

adrianreber commented:
Update PR to target 1.25. Updated KEP PR available at: https://github.com/kubernetes/enhancements/pull/3264

@mrunalp @mikebrow @derekwaynecarr PTAL

adrianreber commented:
> Is there a problem with merging this feature to 1.24?

During today's sig-node meeting there was some feedback that it has been re-targeted to the next release (1.25).

adrianreber commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel

adrianreber commented:
/test pull-kubernetes-integration

adrianreber commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

rphillips commented:
/lgtm

dims commented:
/milestone v1.25

mrunalp commented on pkg/kubelet/kubelet.go:
```diff
@@ -2431,6 +2437,54 @@ func (kl *Kubelet) fastStatusUpdateOnce() {
+			checkpointName := fmt.Sprintf("checkpoint-%s-%s-%d.tar", podFullName, containerName, counter)
+			if _, err := os.Stat(filepath.Join(kl.getCheckpointsDir(), checkpointName)); err == nil {
+				// checkpoint already exists, try another name
+				counter += 1
```
Can we use a timestamp instead of a counter here?

mrunalp commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1514,3 +1519,38 @@ message ReopenContainerLogRequest {
+// Common options used for checkpointing and restoring.
+message CheckpointRestoreOptions {
+    // Location of the checkpoint archive used for export/import
+    string archive_location = 1;
```
Could we shorten this to `location`?

adrianreber replied:
done

mrunalp commented on test/e2e_node/checkpoint_container.go:
```diff
@@ -0,0 +1,157 @@
+		}
+
+		framework.ExpectNoError(err)
+		// TODO: once a container engine implents the Checkpoint CRI API this needs
```
typo: implents -> implements

adrianreber commented:
@mrunalp @rphillips please take another look. Should be less code now than before and the review comments from @mrunalp should be fixed.

CC: @mikebrow the Restore RPC call is gone, you also mentioned in one of the reviews that you would have preferred to have this PR without it.

adrianreber commented:
Rebased to solve merge conflicts

mrunalp commented on pkg/kubelet/server/auth_test.go:
```diff
@@ -110,6 +110,7 @@ func AuthzTestCases() []AuthzTestCase {
 	testPaths := map[string]string{
 		"/attach/{podNamespace}/{podID}/{containerName}":       "proxy",
 		"/attach/{podNamespace}/{podID}/{uid}/{containerName}": "proxy",
+		"/checkpoint/{podNamespace}/{podName}/{containerName}": "proxy",
```
`{podName}` -> `{podID}`

adrianreber replied:
changed

adrianreber commented on pkg/kubelet/server/server.go:
```diff
@@ -403,6 +405,17 @@ func (s *Server) InstallDefaultHandlers() {
+		s.addMetricsBucketMatcher("checkpoint")
+		ws = &restful.WebService{}
+		ws.Path("/checkpoint").Produces(restful.MIME_JSON)
+		ws.Route(ws.POST("/{podNamespace}/{podName}/{containerName}").
```
changed

adrianreber commented:
/test pull-kubernetes-e2e-kind-ipv6

mrunalp commented on pkg/kubelet/server/server.go:
```diff
@@ -878,6 +891,57 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+	if err := s.host.CheckpointContainer(pod.UID, kubecontainer.GetPodFullName(pod), containerName, options); err != nil {
+		response.WriteError(
+			http.StatusInternalServerError,
+			fmt.Errorf("checkpointing of %v failed (%v)", containerName, err),
```
Include the pod namespace/id as well

adrianreber replied:
added pod namespace and id

adrianreber commented:
Applied review changes, rebased and force pushed.

adrianreber commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel

mrunalp commented:
@mikebrow ptal

endocrimes commented on test/e2e_node/checkpoint_container.go:
```diff
@@ -0,0 +1,157 @@
+
+var _ = SIGDescribe("Checkpoint Container [NodeFeature:CheckpointContainer]", func() {
+	f := framework.NewDefaultFramework("checkpoint-container-test")
+	var podClient *framework.PodClient
+	ginkgo.BeforeEach(func() {
+		podClient = f.PodClient()
+	})
```
It's easier to read and maintain test code that doesn't do this, and instead calls `f.PodClient()` where it's needed, or assigns `podClient` local to the `It`.

adrianreber replied:
Thanks. Fixed.

endocrimes commented on pkg/kubelet/kubelet.go:
```diff
@@ -2431,6 +2437,47 @@ func (kl *Kubelet) fastStatusUpdateOnce() {
+// cannot be found an error is returned. If the container is found the container
+// engine will be asked to checkpoint the given container into the kubelet's default
+// checkpoint directory.
+func (kl *Kubelet) CheckpointContainer(
```
Missing unit test? - Should be relatively easy to add one though.

adrianreber replied:
Unit test added.

adrianreber commented:
@endocrimes thanks for the review. I changed the e2e test according to your suggestion and added a unit test. CI seems happy with the changes. PTAL.

endocrimes commented on pkg/kubelet/kubelet.go:
```diff
@@ -2431,6 +2437,47 @@ func (kl *Kubelet) fastStatusUpdateOnce() {
+	}
+
+	if strings.HasPrefix(options.Location, "/") {
+		return fmt.Errorf("archive_location cannot start with '/' but %s was provided", options.Location)
```
```suggestion
		return fmt.Errorf("archive_location cannot start with '/' but %q was provided", options.Location)
```

adrianreber replied:
Thanks, changed.

endocrimes commented on pkg/kubelet/kubelet.go:
```diff
@@ -2431,6 +2437,47 @@ func (kl *Kubelet) fastStatusUpdateOnce() {
+		return fmt.Errorf("archive_location cannot start with '/' but %s was provided", options.Location)
+	}
+
+	if !strings.Contains(options.Location, "/") && options.Location != "" {
```
Missing a test case that covers the negative case of this path - not quite sure what the intent of the behavior is though.

adrianreber replied:
Added an additional test case. That code path will be the main code path. Saving the checkpoint image in the local registry for easy distribution to other registries.

endocrimes reviewed (commented):
Quick skim of tests is looking ok, I'll do a deeper review later.

One maintainability nit: please prefer require over assert. Require ends tests when they fail unlike assert, and makes triaging broken tests in the future a little easier.

adrianreber commented:
> Quick skim of tests is looking ok, I'll do a deeper review later.

Thanks a lot.

> One maintainability nit: please prefer require over assert. Require ends tests when they fail unlike assert, and makes triaging broken tests in the future a little easier.

I changed the unit test to use require everywhere instead of assert.

endocrimes reviewed (commented):
New tests look good. I don't like commenting on style, but there's no need to hard-wrap function calls in tests tho - most of these would be easier to scan on a single line.

Feature overall also lgtm.

endocrimes commented:
/assign @mikebrow
for CRI review

adrianreber commented on pkg/kubelet/server/server_test.go:
```diff
@@ -840,6 +853,78 @@ func TestContainerLogsWithInvalidTail(t *testing.T) {
 	}
 }
 
+func TestCheckpointContainer(t *testing.T) {
```
Ha, I initially thought about doing something likes this but was not sure. I changed the PR to have a table for the checkpoint tests.

adrianreber commented:
@endocrimes I rewrote the new unit test using a table as suggested by you.

mikebrow reviewed (commented):
See timeout and arbitrary path for checkpoint location issues.

adrianreber commented:
/test pull-kubernetes-unit

mikebrow commented on pkg/kubelet/kubelet.go:
```diff
@@ -2436,6 +2442,47 @@ func (kl *Kubelet) fastStatusUpdateOnce() {
+
+	if !strings.Contains(options.Location, "/") && options.Location != "" {
+		options.Location = filepath.Join(kl.getCheckpointsDir(), options.Location)
+	}
```
suggest limiting params to the web service to.. just container id and possibly timeout if others agree timeout should be an option

adrianreber replied:
Filtered out `..`. Over the time developing this PR the option to store the checkpoint as an image in the local registry has become the preferred use case as it makes restoring via create/start possible without an additional CRI RPC call for restore. It also makes it easy to transfer the checkpoint image between Kubernetes Nodes by just pushing the image to registry.

containerd already supports the checkpoint image in my corresponding PR (https://github.com/containerd/containerd/pull/6965) and my CRI-O PR will also support it. I would prefer to keep it the way it is now. Storing the image in a registry provides the option to restore containers in Kubernetes without any additional Kubernetes changes. It can be handled transparently in the underlying CRI implementation.

adrianreber replied:
After discussing this on slack I removed the parameter to specify a custom location and the PR does now only what was agreed upon in the corresponding KEP.

adrianreber commented on pkg/kubelet/cri/remote/remote_runtime.go:
```diff
@@ -1149,3 +1149,42 @@ func (r *remoteRuntimeService) ReopenContainerLog(containerID string) (err error
+	if !r.useV1API() {
+		return errors.New("CheckpointContainer is only supported in the CRI v1 runtime API")
+	}
+	ctx, cancel := getContextWithTimeout(r.timeout)
```
Added a timeout parameter which is passed to the CRI server implementation.

adrianreber commented on pkg/kubelet/cri/remote/remote_runtime.go:
```diff
@@ -1149,3 +1149,42 @@ func (r *remoteRuntimeService) ReopenContainerLog(containerID string) (err error
+	klog.V(10).InfoS(
+		"[RemoteRuntimeService] CheckpointContainer",
+		"options",
+		options,
```
Timeout is part of options now.

adrianreber commented:
/retest-required

mikebrow commented on staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/api.proto:
```diff
@@ -1514,3 +1517,14 @@ message ReopenContainerLogRequest {
+    string container_id = 1;
+    // Location of the checkpoint archive used for export
+    string location = 2;
+    // Timeout after which the kubelet will assume this request has failed
```
```suggestion
    // Timeout in seconds for the checkpoint to complete. Default: 0 (no timeout - run to completion)
```
IMO the default should be either 0 - run to completion or 0 should not be supported at all, make the caller use 1sec or greater timeout. Not sure if using default cri timeout fits the checkpoint use case.  @mrunalp thoughts?

adrianreber replied:
Sounds good to me. I made the change.

mikebrow reviewed (commented):
LGTM per current KEP description.. and would like to see security for this be very tight though I recognize that is not in the KEP.

adrianreber commented:
I see a LGTM from @mikebrow, @rphillips, @endocrimes . The feedback I got from @mrunalp also sounds like a LGTM.

@derekwaynecarr PTAL and approve if appropriate.

mrunalp commented:
@tallclair ptal

adrianreber commented:
In today's sig-node meeting the security implications of this newly added kubelet API were discussed again. Just to make sure we do not add an interface that can be misused.

My recollection of the discussion was that any access to the kubelet API requires authorization. Whenever I test it I need to add `--cert /var/run/kubernetes/client-admin.crt --key /var/run/kubernetes/client-admin.key` to the curl command-line.
We are creating the checkpoint archive with `0600` so that only the local root user can access the checkpoint. I think the consensus was that this works for Alpha. For Beta there was the suggestion to have additional authorization for this newly introduced API endpoint to the kubelet. I am not sure if any other kubelet API currently has additional authorization requirements.

If I remember it correctly @tallclair was part of that initial discussion about adding additional authorization to the kubelet `/checkpoint/` API and Beta was the point where we would look into that. I hope there are all already endpoints in the kubelet with additional authorization requirements.

mikebrow commented:
| We are creating the checkpoint archive with 0600 so that only the local root user can access the checkpoint.

and over here https://github.com/opencontainers/runc/pull/2139/files#r335825579

A couple questions/thoughts on the security aspect.

Should the same user Authorized to enable checkpointing be allowed to make the checkpoint request?  
Should the checkpoint be on a unique port to separate monitoring access from checkpoint access? 
Should the root user be considered, in all cases, a user authorized to access the contents of the node's containers?  
Should there be support/requirements for using fs level encryption when checkpointing encrypted containers?

mrunalp commented on pkg/kubelet/cri/remote/remote_runtime.go:
```diff
@@ -1149,3 +1149,50 @@ func (r *remoteRuntimeService) ReopenContainerLog(containerID string) (err error
+			return getContextWithTimeout(time.Duration(options.Timeout) * time.Second)
+		}
+		// No timeout requested means we fall back to no time out.
+		return getContextWithCancel()
```
I think we should default to some value when none is provided.

adrianreber replied:
I am happy to implement whatever is seen as useful. The current no timeout implementation was based on @mikebrow suggestion.

mikebrow replied:
Or you could throw a zero timeout value is not supported error?

mikebrow replied:
https://github.com/kubernetes/kubernetes/pull/104907#discussion_r900821827

mrunalp replied:
Yeah, we can error on negative value here :+1:

adrianreber replied:
So, after chatting with @mrunalp and @mikebrow I changed it so that `0` means use CRI default and anything `>0` means use that user-specific timeout. Any negative values are an error.

adrianreber commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

mrunalp commented on pkg/kubelet/server/server.go:
```diff
@@ -878,6 +891,63 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+// podNamespace, pod and container actually exist and only then calls out
+// to the runtime to actually checkpoint the container.
+func (s *Server) checkpoint(request *restful.Request, response *restful.Response) {
+	if request == nil || response == nil {
```
Is this check needed? I don't see it in the other handlers.

adrianreber replied:
Just wanted to be on the safe side. Removed the check now.

mrunalp commented on pkg/kubelet/server/server.go:
```diff
@@ -878,6 +891,63 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+		return
+	}
+	options := &runtimeapi.CheckpointContainerRequest{}
+	if request.Request.ContentLength > 0 {
```
Not sure if this is needed either. Can we make the parameter extraction similar to the other handlers for consistency?

adrianreber replied:
I switched to query parameters as part of the URL as seen in exec, run, tail.

mrunalp commented on pkg/kubelet/cri/remote/remote_runtime.go:
```diff
@@ -1149,3 +1149,55 @@ func (r *remoteRuntimeService) ReopenContainerLog(containerID string) (err error
+	}
+
+	ctx, cancel := func() (context.Context, context.CancelFunc) {
+		if options.Timeout == 0 {
```
Maybe if it is less than the CRI configured value, we set it to the CRI value?

adrianreber replied:
changed

mrunalp commented on pkg/kubelet/container/runtime.go:
```diff
@@ -119,6 +119,9 @@ type Runtime interface {
 	// CIDR value down to the runtime shim.
 	UpdatePodCIDR(podCIDR string) error
+	// CheckpointContainer tells the runtime to checkpoint a container
+	// and store the resulting archive to checkpointDir
```
nit: checkpointDir -> checkpoint directory

adrianreber replied:
changed

mrunalp commented:
Left couple of nits, otherwise looking good.

adrianreber commented:
> Left couple of nits, otherwise looking good.

Changed as suggested.

derekwaynecarr commented on pkg/kubelet/cri/remote/remote_runtime.go:
```diff
@@ -1149,3 +1149,58 @@ func (r *remoteRuntimeService) ReopenContainerLog(containerID string) (err error
+
+	ctx, cancel := func() (context.Context, context.CancelFunc) {
+		defaultTimeout := int64(r.timeout / time.Second)
+		if options.Timeout > defaultTimeout {
```
what could cause this timeout to need to vary?

adrianreber replied:
Checkpoint duration directly depends on the number of memory pages the processes in the container require. For very large containers checkpointing can take longer than the default CRI timeout of 2 seconds.

derekwaynecarr commented on pkg/kubelet/kubelet.go:
```diff
@@ -1251,6 +1254,9 @@ func (kl *Kubelet) setupDataDirs() error {
 	if err := os.MkdirAll(kl.getPodResourcesDir(), 0750); err != nil {
 		return fmt.Errorf("error creating podresources directory: %v", err)
 	}
+	if err := os.MkdirAll(kl.getCheckpointsDir(), 0750); err != nil {
```
we should only create the checkpoint directory if the feature gate is enabled.

adrianreber replied:
Good point. Changed.

derekwaynecarr commented on pkg/kubelet/server/server.go:
```diff
@@ -878,6 +891,69 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+	containerName := request.PathParameter("containerName")
+
+	found := false
+	for _, container := range pod.Spec.Containers {
```
this means we are unable to checkpoint init containers or ephemeral containers.

can you add a check for `pod.Spec.InitContainers` and a check for `pod.Spec.EphemeralContainers` if the feature gate for ephemeral containers is enabled?

adrianreber replied:
Thanks for catching this. I changed it and was able to checkpoint init and ephemeral containers.

derekwaynecarr reviewed (changes requested):
@adrianreber this is super close.

can you address the following:

1) ensure checkpoint directory is only created if the feature gate is enabled
2) ensure i can checkpoint an init container and/or ephemeral container

Thanks!

adrianreber commented:
> @adrianreber this is super close.

:fireworks: 

> can you address the following:
> 
>     1. ensure checkpoint directory is only created if the feature gate is enabled

Done. Good point.

>     2. ensure i can checkpoint an init container and/or ephemeral container

Also done. Thanks for your very specific review. That made it really easy to implement the requested changes.

derekwaynecarr commented:
Thanks for the updates as requested.

/approve
/lgtm

adrianreber commented:
/test pull-kubernetes-e2e-kind-ipv6
/test pull-kubernetes-e2e-kind

adrianreber commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

adrianreber commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

adrianreber commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd
